Fix Keyboard Screen Tests

CananArikan opened the pull request:
closes #466 
The parent issue to https://github.com/willowtreeapps/vocable-ios/issues/466 is https://github.com/willowtreeapps/vocable-ios/issues/405

# Description of Work
Fixed broken keyboard tests. This is part of the effort to unlink the chained tests.

CananArikan committed:
Merge branch 'develop' into test/466-fix-keyboard-screen-tests

CananArikan committed:
Merge branch 'develop' into test/466-fix-keyboard-screen-tests

CananArikan committed:
Fixed Keyboard Screen Tests

testKeyboardOutputIsDisplayed is fixed. testAddPhraseToMySayingsFromKeyboard and testRemovePhraseFromMySayingsFromKeyboard are added.

CananArikan committed:
Added Helper Functions to Main Screen

Clstroud commented on VocableUITests/Tests/KeyboardScreenTests.swift:
```diff
@@ -3,30 +3,59 @@
+         
+        XCTAssertFalse(mainScreen.isTextDisplayed(testPhrase), "Expected the phrase \(testPhrase) to be deleted from 'My Sayings'")
+    }
+    
+
+    
```
```suggestion
```

CananArikan replied:
Fixed.

Clstroud commented on VocableUITests/Screens/MainScreen.swift:
```diff
@@ -51,4 +51,39 @@ class MainScreen: BaseScreen {
+        }
+    }
+    
+    func findPhrase(phrase: String) {
+        
+        let predicate = NSPredicate(format: "label CONTAINS %@", phrase)
+        
+        // Loop through each custom category page to find our phrase
+        for _ in 1...totalPageCount {
+            if app.cells.staticTexts.containing(predicate).element.exists {
+                break
+            } else {
+                paginationRightButton.tap()
+            }
+        }
+    }
```
I like the structure y'all have come up with here! I've got one overarching suggestion we can discuss, but I want to emphasize that this is just me thinking aloud 😄
 
I think you might be able to find some harmony with the `XCUITest` framework by separating the notion of _locating_ a particular `XCUIElement` from _interacting_ with that element. Currently there is an inconsistency where one function specifies that it will find a category and also tap, but the other finds a phrase and doesn't really let the caller know if it was successful or not 🤭 

So in practice I might suggest something like (I'm making this up on the fly):
`func locateCategoryCell(matching identifier: String) -> XCUIElement?` and
`func locatePhraseCell(withText text: String) -> XCUIElement?`
You would be able to very cleanly add an assertion that your element exists (is non-nil) and then tap it like you would any other element.

You could even compose the two with a new function:
`func locatePhraseCell(withText text: String, inCategoryMatchingIdentifier categoryID: String) -> XCUIElement?`
in order to cut down on future boilerplate.

CananArikan replied:
Functions are updated as suggested.

Clstroud commented on VocableUITests/Screens/MainScreen.swift:
```diff
@@ -51,4 +51,39 @@ class MainScreen: BaseScreen {
     }
+    
+    func scrollToMySayings() {
+        let mySayingsIdentifier = "category_title_cell_preset_user_favorites"
```
It may make sense to have an enum of Categories where these identifiers can be stored. Then you could refer to those identifiers when scrolling to a particular category instead of having to make individual functions for each 🤔

rhsalas3 replied:
That's something I've been thinking about as well, while we start adding more and more identifiers. And especially now that we know that categories already have unique IDs we can reference.

To further clarify your suggestion when it comes to this particular function, does that look something like `func scrollToCategory(category: Category)` thus a call to it would be `mainScreen.scrollToCategory(Category.MySayings)`?

CananArikan replied:
Added CategoryIdentifier/CategoryTitleCellIdentifier structs so that we now have a type to be used for passing along the category title cell element identifiers.

rowildcard commented on VocableUITests/Tests/KeyboardScreenTests.swift:
```diff
@@ -3,30 +3,59 @@
+        mainScreen.scrollToMySayings()
+        mainScreen.findPhrase(phrase: testPhrase)
+        
+        XCTAssertTrue(mainScreen.isTextDisplayed(testPhrase), "Expected the phrase \(testPhrase) to be added to an displayed in 'My Sayings'")
```
Typo. I believe you wanted your error message to say, "... added to _and_ displayed..."

CananArikan replied:
Fixed.

rowildcard commented on VocableUITests/Tests/KeyboardScreenTests.swift:
```diff
@@ -3,30 +3,59 @@
+        mainScreen.scrollToMySayings()
+        mainScreen.findPhrase(phrase: testPhrase)
+       
+        XCTAssertTrue(mainScreen.isTextDisplayed(testPhrase), "Expected the phrase \(testPhrase) to be added to an displayed in 'My Sayings'")
```
Ditto.
Typo. I believe you wanted your error message to say, "... added to _and_ displayed..."

CananArikan replied:
Fixed.

CananArikan committed:
Merge branch 'develop' into test/466-fix-keyboard-screen-tests

CananArikan committed:
PR Fixes

Added CategoryIdentifier/CategoryTitleCellIdentifier structs so that we now have a type to be used for passing along the category title cell element identifiers.
Launch argument to reset app data is implemented. No longer need random strings.

CananArikan committed:
Merge branch 'develop' of https://github.com/willowtreeapps/vocable-ios into test/466-fix-keyboard-screen-tests

Clstroud commented on VocableUITests/Tests/CategoryIdentifier.swift:
```diff
@@ -0,0 +1,48 @@
+    private let categoryIdentifier: CategoryIdentifier
+    let categoryTitleCellPrefix: String = "category_title_cell_"
+    
+    // Empty: used to retrieve the prefix only
+    init() {
+        self.categoryIdentifier = CategoryIdentifier("")
+    }
```
I think you might be able to declare the prefix as `static let prefix: String = "category_title_cell_"`, which would mean you don't have to actually instantiate this struct to access it: `CategoryTitleCellIdentifier.prefix`

CananArikan replied:
Fixed.

Clstroud commented on VocableUITests/Screens/MainScreen.swift:
```diff
@@ -25,6 +25,22 @@ class MainScreen: BaseScreen {
+        // Build our query that first finds all category title cells, then finds among those the one that is selected
+        let query = XCUIApplication().cells.containing(identifierPredicate).containing(isSelectedPredicate)
+        
+        // Remove the identifierPrefix
+        let categoryTitleCellIdentifierString = query.element.identifier.replacingOccurrences(of: identifierPrefix, with: "")
+          
+        // Return the selected category's full identifier
+        return CategoryTitleCellIdentifier(CategoryIdentifier(categoryTitleCellIdentifierString))
```
I think it might make sense to move this logic into `CategoryTitleCellIdentifier` itself, giving it a [failable initializer](https://www.hackingwithswift.com/sixty/10/9/failable-initializers) that takes a `String`, and attempts to locate and remove the prefix.

CananArikan replied:
Updated as suggested.

Clstroud reviewed (commented):
I have just two small suggestions and then I think this will be 💯 Y'all are doing great

CananArikan committed:
Refactored CategoryTitleCellIdentifier

Refactored CategoryTitleCellIdentifier to initialize with a known String identifier, which we used for selectedCategoryCell.

Clstroud reviewed (approved):
Nicely done! 🚢 💯

CananArikan committed:
Fixed typo
